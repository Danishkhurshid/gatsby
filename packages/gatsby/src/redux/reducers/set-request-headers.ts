import reporter from "gatsby-cli/lib/reporter"
import url from "url"

import { IGatsbyState, ISetDomainRequestHeaders } from "../types"

export const setRequestHeadersReducer = (
  state: IGatsbyState["requestHeaders"] = new Map(),
  action: ISetDomainRequestHeaders
): IGatsbyState["requestHeaders"] => {
  switch (action.type) {
    case `SET_REQUEST_HEADERS`: {
      const { headers, domain, pluginName } = action.payload

      const noHeaders = typeof headers !== `object`
      const noDomain = typeof domain !== `string`

      if (noHeaders) {
        reporter.warn(
          `Plugin ${pluginName} called actions.setRequestHeaders with a headers property that isn't an object.`
        )
      }

      if (noDomain) {
        reporter.warn(
          `Plugin ${pluginName} called actions.setRequestHeaders with a domain property that isn't a string.`
        )
      }

      if (noDomain || noHeaders) {
        return state
      }

      const baseDomain = url.parse(domain)?.hostname

      if (baseDomain) {
        state.set(baseDomain, headers)
      } else {
        reporter.warn(
          `Plugin ${pluginName} called actions.setRequestHeaders with a domain that is not a valid URL. (${domain})`
        )
      }

      return state
    }
    default:
      return state
  }
}
