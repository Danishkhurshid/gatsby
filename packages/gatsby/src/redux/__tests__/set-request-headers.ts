import { setRequestHeadersReducer as reducer } from "../reducers/set-request-headers"
import { actions } from "../actions"
import * as reporter from "gatsby-cli/lib/reporter"

jest.mock(`gatsby-cli/lib/reporter`, () => {
  return {
    __esModule: true,
    warn: jest.fn(),
  }
})

const testPlugin = {
  name: `gatsby-source-test`,
}

describe(`Add request headers`, () => {
  it(`allows you to add request headers`, () => {
    const action = actions.setRequestHeaders(
      {
        domain: `https://testdomain.com/subpath`,
        headers: {
          "X-Header": `test`,
        },
      },
      testPlugin
    )
    expect(action.payload.domain).toBe(`testdomain.com`)
    expect(action.payload.headers).toEqual({
      "X-Header": `test`,
    })
    expect(action).toMatchSnapshot()

    const state = reducer(undefined, action)

    expect(state.get(`testdomain.com`)).toEqual({
      "X-Header": `test`,
    })
    expect(state).toMatchSnapshot()
  })

  it(`fails if domain is missing`, () => {
    actions.setRequestHeaders(
      {
        headers: {
          "X-Header": `test`,
        },
      },
      testPlugin
    )

    expect(reporter.warn).toHaveBeenCalledWith(
      `Plugin gatsby-source-test called actions.setRequestHeaders with a domain property that isn't a string.`
    )
  })

  it(`fails if headers are missing`, () => {
    actions.setRequestHeaders(
      {
        domain: `https://testdomain.com/subpath`,
      },
      testPlugin
    )

    expect(reporter.warn).toHaveBeenCalledWith(
      `Plugin gatsby-source-test called actions.setRequestHeaders with a headers property that isn't an object.`
    )
  })
})
