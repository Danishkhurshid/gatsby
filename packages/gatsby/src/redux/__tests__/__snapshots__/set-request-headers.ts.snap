// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add request headers allows you to add request headers 1`] = `
Object {
  "payload": Object {
    "domain": "testdomain.com",
    "headers": Object {
      "X-Header": "test",
    },
  },
  "type": "SET_REQUEST_HEADERS",
}
`;

exports[`Add request headers allows you to add request headers 2`] = `
Map {
  "testdomain.com" => Object {
    "X-Header": "test",
  },
}
`;
